File: 02 - BettingService/src/services/betting_service.py
from typing import List, Optional
from src.models.betting import (
    Match, BettingOdds, Bet, BetType, PlaceBetRequest, 
    BettingOddsResponse, BetResponse
)
from src.storage.memory_storage import storage


class BettingService:
    def __init__(self):
        self.storage = storage

    def get_all_matches(self) -> List[Match]:
        """Get all available matches"""
        return self.storage.get_all_matches()

    def get_match_odds(self, match_id: str) -> Optional[BettingOddsResponse]:
        """Get all odds for a specific match"""
        match = self.storage.get_match(match_id)
        if not match:
            return None
        
        odds = self.storage.get_odds_for_match(match_id)
        return BettingOddsResponse(match=match, odds=odds)

    def get_odds_by_type(self, match_id: str, bet_type: BetType) -> List[BettingOdds]:
        """Get odds for a specific bet type on a match"""
        return self.storage.get_odds_by_type(match_id, bet_type)

    def place_bet(self, bet_request: PlaceBetRequest) -> BetResponse:
        """Place a new bet"""
        # Validate match exists
        match = self.storage.get_match(bet_request.match_id)
        if not match:
            return BetResponse(
                success=False,
                message=f"Match with ID {bet_request.match_id} not found"
            )

        # Validate odds exist for this bet
        odds = self.storage.get_specific_odds(
            bet_request.match_id, 
            bet_request.bet_type, 
            bet_request.option
        )
        if not odds:
            return BetResponse(
                success=False,
                message=f"No odds available for {bet_request.bet_type.value} with option '{bet_request.option}'"
            )

        # Create the bet
        potential_win = bet_request.stake * odds.odds
        bet = Bet(
            match_id=bet_request.match_id,
            bet_type=bet_request.bet_type,
            option=bet_request.option,
            stake=bet_request.stake,
            odds=odds.odds,
            potential_win=potential_win
        )

        # Store the bet
        stored_bet = self.storage.place_bet(bet)
        
        return BetResponse(
            success=True,
            bet=stored_bet,
            message="Bet placed successfully"
        )

    def get_bet(self, bet_id: str) -> Optional[Bet]:
        """Get a specific bet by ID"""
        return self.storage.get_bet(bet_id)

    def get_all_bets(self) -> List[Bet]:
        """Get all placed bets"""
        return self.storage.get_all_bets()

    def get_bets_for_match(self, match_id: str) -> List[Bet]:
        """Get all bets for a specific match"""
        return self.storage.get_bets_for_match(match_id)


# Global service instance
betting_service = BettingService()

File: 06 - Frontend/src/services/bettingService.ts
import { bettingApi } from './api';
import { BettingOdds, Match, ApiResponse } from '../types';

export const bettingService = {
  // Get all available matches for betting
  getMatches: async (): Promise<Match[]> => {
    try {
      const response = await bettingApi.get<ApiResponse<Match[]>>('/api/matches');
      
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
      
      return [];
    } catch (error) {
      console.error('Error fetching betting matches:', error);
      throw error;
    }
  },

  // Get odds for a specific match
  getMatchOdds: async (matchId: string): Promise<BettingOdds[]> => {
    try {
      const response = await bettingApi.get<ApiResponse<BettingOdds[]>>(`/api/matches/${matchId}/odds`);
      
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
      
      return [];
    } catch (error) {
      console.error('Error fetching match odds:', error);
      throw error;
    }
  },

  // Get odds for a specific bet type
  getBetTypeOdds: async (matchId: string, betType: string): Promise<BettingOdds | null> => {
    try {
      const response = await bettingApi.get<ApiResponse<BettingOdds>>(`/api/matches/${matchId}/odds/${betType}`);
      
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
      
      return null;
    } catch (error) {
      console.error('Error fetching bet type odds:', error);
      return null;
    }
  },

  // Place a bet
  placeBet: async (betData: {
    match_id: string;
    bet_type: string;
    option: string;
    stake: number;
  }): Promise<any> => {
    try {
      const response = await bettingApi.post<ApiResponse<any>>('/api/bets', betData);
      
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
      
      throw new Error(response.data.message || 'Failed to place bet');
    } catch (error) {
      console.error('Error placing bet:', error);
      throw error;
    }
  },

  // Get all bets
  getAllBets: async (): Promise<any[]> => {
    try {
      const response = await bettingApi.get<ApiResponse<any[]>>('/api/bets');
      
      if (response.data.success && response.data.data) {
        return response.data.data;
      }
      
      return [];
    } catch (error) {
      console.error('Error fetching bets:', error);
      throw error;
    }
  },

  // Health check
  healthCheck: async (): Promise<boolean> => {
    try {
      const response = await bettingApi.get('/api/health');
      return response.status === 200;
    } catch (error) {
      console.error('Betting Service health check failed:', error);
      return false;
    }
  },
};